### PR TITLE
HTML encoding fix tool

### PR DESCRIPTION
**Pull request by btry**
CLI tool to fix data having bad HTML encoding in database. 

Currently fixes double encoded items 

internal ref 24943

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | internal: 24943

### CODE REVIEW

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+     */
+    const ERROR_UPDATE_FAILED = 4;
+
+    protected function configure()
+    {
+        parent::configure();
+
+        $this->setName('glpi:database:fix_html_encoding');
+        $this->setAliases(['db:fix_html']);
+        $this->setDescription(__('Enable timezones usage.'));
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:80]: Should be changed.

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            'field',
+            null,
+            InputOption::VALUE_IS_ARRAY | InputOption::VALUE_REQUIRED,
+            __('Field of the item to fix')
+            . PHP_EOL
+            . __('"--field content" will fix the field "content"')
+            . PHP_EOL
+        );
+
+        $this->addUsage('--field name --field content');
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:112]: Usage example should contains all required options. e.g. `--itemtype=ITILFollowup --id=42 --field=content`.

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    {
+        global $DB;
+
+        $itemtype = $input->getOption('itemtype');
+        if (empty($itemtype) || !class_exists($itemtype) || !is_a($itemtype, CommonDBTM::class, true)) {
+            $this->output->writeln(
+                '<error>' . sprintf(__('Itemtype %s not found'), $itemtype) . '</error>',
+                OutputInterface::VERBOSITY_QUIET
+            );
+            $this->outputSessionBufferedMessages([WARNING, ERROR]);
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:125]: This is not usefull here. Same for all other occurences.
```suggestion
```

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                $this->outputSessionBufferedMessages([WARNING]);
+                return self::ERROR_UPDATE_FAILED;
+            }
+        }
+
+        return 0;
+    }
+
+    /**
+     * Remove double encoding og HTML tags
```
Comment (btry) [inline src/Console/Database/FixHtmlEncodingCommand.php:196]:
```suggestion
     * Remove double encoding of HTML tags
```

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        }
+
+        return 0;
+    }
+
+    /**
+     * Remove double encoding of HTML tags
+     * character < is encoded &#38;lt; but should  be encoded &#60;
+     * character > is encoded &#38;gt; but should  be encoded &#62;
```
Comment (btry) [inline src/Console/Database/FixHtmlEncodingCommand.php:198]:
```suggestion
     * character < is encoded &#38;lt; but should be encoded &#60;
     * character > is encoded &#38;gt; but should be encoded &#62;
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 6]
```diff
@@ -100,7 +100,14 @@ protected function configure()
             __('Field of the item to fix')
         );
 
-        $this->addUsage('--itemtype=ITILFollowup --id=42 --field=content');
+        $this->addOption(
+            'dump',
+            null,
+            InputOption::VALUE_OPTIONAL,
+            __('Field of the item to fix')
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:107]:
```suggestion
            __('Path of file containing dump of existing values.')
```

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 7]
```diff
@@ -157,22 +164,63 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        if ($input->getOption('dump')) {
+            $dump_file_name = $input->getOption('dump');
+            if (file_put_contents($dump_file_name, $dump_content) == strlen($dump_content)) {
+                $this->output->writeln(
+                    '<comment>' . sprintf(__('File %s contains SQL queries that can be used to rollback command.'), $dump_file_name) . '</comment>',
+                    OutputInterface::VERBOSITY_QUIET
+                );
+            } else {
+                $this->output->writeln(
+                    '<comment>' . sprintf(__('Failed to write rollback SQL queries file %s'), $dump_file_name) . '</comment>',
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:221]:
```suggestion
                    '<comment>' . sprintf(__('Failed to write rollback SQL queries file %s.'), $dump_file_name) . '</comment>',
```

**Review by cedric-anne — APPROVED** [review]

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            );
+            foreach ($failed_items as $item_id) {
+                $this->output->writeln(
+                    '<error>' . sprintf(__('Itemtype %s ID %s'), $itemtype, $item_id) . '</error>',
+                    OutputInterface::VERBOSITY_QUIET
+                );
+            }
+            return self::ERROR_UPDATE_FAILED;
+        }
+
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:170]: You should add a success message, like:
```suggestion

        $output->writeln('<info>' . __('HTML encoding has been fixed.') . '</info>');
```

File: src/Console/Database/FixHtmlEncodingCommand.php [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+     */
+    const ERROR_ROLLBACK_FILE_FAILED = 5;
+
+    protected function configure()
+    {
+        parent::configure();
+
+        $this->setName('glpi:database:fix_html_encoding');
+        $this->setAliases(['db:fix_html']);
+        $this->setDescription(__('Fix Html encoding in database.'));
```
Comment (cedric-anne) [inline src/Console/Database/FixHtmlEncodingCommand.php:87]:
```suggestion
        $this->setDescription(__('Fix HTML encoding in database.'));
```

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**